docs/overview.adoc

```diff
@@ -0,0 +1,103 @@
+That means that filters generally don't know what other filters might be present in the chain, and what they might be doing to messages.
+When a filter forwards a request or response it doesn't know whether the message be being sent to the next filter in the chain, or straight back to the client.
+
+Such composition is important because it means a _proxy user_ can configure multiple filters (possibly written by several _filter authors_) and expect to get the combined effect of all of them.
+
+It's never quite that simple, of course.
+In practice they will often need to understand what each filter does in some detail in order to be able to operate their proxy properly, for example by understanding whatever metrics each filter is emitting.
+
+== Implementation
+
+The proxy is written in Java, on top of https://netty.io[Netty].
+The usual https://netty.io/4.1/api/io/netty/channel/ChannelHandler.html[`ChannelHandlers`] provided by the Netty project are used where appropriate (e.g. SSL support uses https://netty.io/4.1/api/io/netty/handler/ssl/SslHandler.html[`SslHandler`]), and Kroxy provides Kafka-specific handlers of its own.
+
+The Kafka-aware parts use the Apache Kafka project's own classes for serialization and deserialization.
+
+Protocol filters get executed using a handler-per-filter model.
+
+== Deployment topologies
+
+The proxy supports a range of possibly deployment topologies.
+Which style is used depends on what the proxy is meant to _achieve_, architecturally speaking.
+Broadly speaking a proxy instance can be deployed:
+
+As a forward proxy::
+proxying the access of one or more clients to a particular cluster/broker that might also accessible (to other clients) directly.
++
+// TODO include a diagram
++
+Topic-level encryption provides one example use case for a forward proxy-style deployment.
+This might be applicable when using clients that don't support interceptors, or if an organisation wants to apply the same encryption policy in a single place, securing access to the keys within their network.
+
+As a reverse proxy::
+proxying access for all clients trying to reach a particular cluster/broker.
++
+// TODO include a diagram
++
+Transparent multi-tenancy provides an example use case for a reverse proxy.
+While Apache Kafka itself has some features that enable multi-tenancy, they rely on topic name prefixing as the primary mechanism for ensuring namespace isolation.
+Tenants have to adhere to the naming policy and know they're a tenant of a larger shared cluster.
++
+_Transparent_ multi-tenancy means each tenant has the illusion of having their own cluster, with almost complete freedom over topic and group naming, while still actually sharing a cluster.
+
+// TODO we probably don't need the level of detail below, just summarize
+// and provide the detail in the deploying section
+
+We can further classify deployment topologies in how many proxy instances are used. For example:
+
+Proxy-per-broker::
```

Comment:
Since this might include being a side-car in some setups, should we explicitly call that out? Since side-car deployment is a very popular model for proxies generally, I expect people to be searching the docs explicitly for that. 
(Happy to be added as a TODO and I can help flesh it out)

Reply:
I see kroxy as a component that will run on bare mental as well as containerised.   I imagine that the kroxy ecosystem will eventually have kube operators, but I don't think we want kube to leak into the core.  WDYT?

Reply:
> I see kroxy as a component that will run on bare mental as well as containerised. I imagine that the kroxy ecosystem will eventually have kube operators, but I don't think we want kube to leak into the core. WDYT?

Oh I didn't mean that Kroxylicious should rely anyhow on Kube. I meant the side-car deployment model generally, whether it ends up as a 2nd container in a Kafka pod or a standalone process in a VM where Kafka is running. I think the side-car model isn't Kube?